src/core_plugins/kibana/public/dashboard/actions/embeddables.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from 'lodash';
+import { Action, Dispatch } from 'redux';
+import { createAction } from 'redux-actions';
+import { CoreKibanaState } from '../../selectors';
+import { PanelId, PanelState } from '../selectors';
+import { updatePanel } from './panels';
+
+import { EmbeddableMetadata, EmbeddableState } from 'ui/embeddable';
+import {
+  getEmbeddableCustomization,
+  getPanel,
+} from '../../selectors/dashboard_selectors';
+
+export type EmbeddableIsInitializingActionPayload = string;
+
+export enum EmbeddableActionTypeKeys {
```

Comment:
cc @timroes - you recall this conversation at all?  Happy to switch, just don't want to embark on the refactor unless I'm sure it's the way we decided to go.

Reply:
If we don't want to have any of those in, we should setup linting for that. I'm pretty sure I already introduced them in a couple of places (though some might not yet be merged). I also don't see why we should not use enums, since we even use 'enums', i.e. what it transpires to in several places in JavaScript code, e.g. https://github.com/elastic/kibana/blob/master/src/ui/public/vis/vis_category.js

Reply:
I'd like to use enums too, I think it's nice to take advantage of all the language has to offer.  But also since it seems like not everyone recalls this debate, I agree we should add the linting rules and maybe make it a little more visible when making this decision.

For the sake of this PR, I'd like to leave in, unless we can come to a relatively quick resolution.

Reply:
> i.e. what it transpires to in several places in JavaScript code, e.g. https://github.com/elastic/kibana/blob/master/src/ui/public/vis/vis_category.js

Well, if transpiled version was like you're refering to nobody would definitely object :)

```js
const CATEGORY = {
  BASIC: 'basic',
  DATA: 'data',
  GRAPHIC: 'graphic',
  MAP: 'map',
  OTHER: 'other',
  TIME: 'time',
  // Hidden is a specific category and doesn't need a display name below
  HIDDEN: 'hidden'
};
```
But [real transpiled version looks a bit different](http://www.typescriptlang.org/play/#src=enum%20CATEGORY%20%7B%0D%0A%20%20BASIC%20%3D%20'basic'%2C%0D%0A%20%20DATA%20%3D%20'data'%2C%0D%0A%20%20GRAPHIC%20%3D%20'graphic'%2C%0D%0A%20%20MAP%20%3D%20'map'%2C%0D%0A%20%20OTHER%20%3D%20'other'%2C%0D%0A%20%20TIME%20%3D%20'time'%2C%0D%0A%20%20%2F%2F%20Hidden%20is%20a%20specific%20category%20and%20doesn't%20need%20a%20display%20name%20below%0D%0A%20%20HIDDEN%20%3D%20'hidden'%0D%0A%7D%3B):
```js
var CATEGORY;
(function (CATEGORY) {
    CATEGORY["BASIC"] = "basic";
    CATEGORY["DATA"] = "data";
    CATEGORY["GRAPHIC"] = "graphic";
    CATEGORY["MAP"] = "map";
    CATEGORY["OTHER"] = "other";
    CATEGORY["TIME"] = "time";
    // Hidden is a specific category and doesn't need a display name below
    CATEGORY["HIDDEN"] = "hidden";
})(CATEGORY || (CATEGORY = {}));
```

Having said that I'm not opposed to the idea of using TS enums, I just want us to carefully pick every TS feature, know exactly what is behind it and what consequences it may have if we decide to drop TS (I hope not) :)

Reply:
We should use TypeScript to its fullest without regard to how any single feature compiles to JavaScript. I understand your concern @azasypkin since you spent a bunch of time cleaning up the compiled JavaScript from a TypeScript codebase, but the experiment with TypeScript is over - we're now committed to rolling this out and using it for the long term, so let's embrace it.

At this point, there are really only two ways I can see us abandoning TypeScript:

1. TS is abandoned
2. TS goes in a direction we're fundamentally unhappy with

I don't anticipate either of those things happening, but in the event that they do, we have options to clean up any gross code. The example you just listed could be codeshifted away, for example. The thing is, the compiled output can and likely will change as we use and upgrade TypeScript. So what is a gross output today might be fine tomorrow, and vice versa. All the more reason not to evaluate this on a feature by feature basis.

Reply:
> the experiment with TypeScript is over - we're now committed to rolling this out and using it for the long term, so let's embrace it.

Thanks @epixa , that's exactly what I wanted to hear!